### DISCUSSION TITLE
warning: initialization discards 'const' qualifier from pointer target type

### DISCUSSION THREAD

**Discussion by sabsaback** [micropython/micropython#11142] [General]
I am trying to export a module with some functions and I compile with meson. This is throwing a warning and I am wondering if there is any way to get rid of it? Am I doing something wrong?

```
 ./../subprojects/micropython/py/obj.h:309:23: warning: initialization discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
   309 | #define MP_ROM_PTR(p) (p)
       |                       ^
 ../test.c:9:38: note: in expansion of macro 'MP_ROM_PTR'
     9 |     { MP_ROM_QSTR(MP_QSTR_test),     MP_ROM_PTR(&test_obj) },
```

My test.c:

```
#include "test.h"
STATIC MP_DEFINE_CONST_FUN_OBJ_1(test_obj, Test_testfunction);
 
STATIC const mp_map_elem_t Test_globals_table[] = {
     // module name
     { MP_ROM_QSTR(MP_QSTR___name__), MP_OBJ_NEW_QSTR(MP_QSTR_Test) },
     // function names
     { MP_ROM_QSTR(MP_QSTR_test),     MP_ROM_PTR(&test_obj) },
};
 
STATIC MP_DEFINE_CONST_DICT (mp_module_Test_globals, Test_globals_table);
 
const mp_obj_module_t Test_cmodule = {	
     .base = { &mp_type_module },
     .globals = (mp_obj_dict_t*)&mp_module_Test_globals,
};
 
MP_REGISTER_MODULE(MP_QSTR_Test, Test_cmodule, 1);
```

My test.h:

```
#ifndef _TEST_H_
#define _TEST_H_
 
#ifdef __cplusplus
extern "C" {
#endif
 
#include "py/obj.h"
 
mp_obj_t Test_testfunction(mp_obj_t o_in);
#ifdef __cplusplus
}
#endif
#endif // _TEST_H_
```

My function is in testimpl.cpp:

```
#include "test.h"
#ifndef NO_QSTR
#endif

extern "C" {
#include "py/builtin.h"
#include "py/objlist.h"
#include "py/runtime.h"
}
 
mp_obj_t Test_testfunction(mp_obj_t o_in) {
    int i = mp_obj_get_int(o_in);
    return mp_obj_new_int(i);
}
```

**Comment by sabsaback** [comment]
This got rid of the warnings, but since I am new at micropython/c /c++ I am not sure what effects this causes: (const mp_obj_t)
```
STATIC const mp_map_elem_t Test_globals_table[] = {
     // module name
     { MP_ROM_QSTR(MP_QSTR___name__), MP_OBJ_NEW_QSTR(MP_QSTR_Test) },
     // function names
     { MP_ROM_QSTR(MP_QSTR_test),     (const mp_obj_t)&test_obj },
};
```

**Comment by dlech** [comment]
I think you want `mp_rom_map_elem_t` instead of `mp_map_elem_t` for your globals table.

**Reply by sabsaback** [reply]
Thank you for your answer. 

Interesting, could you explain the difference? In what cases should I use `mp_map_elem_t` and in what `mp_rom_map_elem_t`

**Reply by dlech** [reply]
Generally, when defining a MicroPython module in C, data structures are usually declared static const (read-only). On microcontrollers, this means the data is stored in flash memory (ROM) and never copied to RAM to leave more RAM free for other things. Trying to write to flash memory usually causes a crash, so we use `const` so that the compiler will prevent us from trying to accidentally write to read-only memory.

**Reply by sabsaback** [reply]
Thank you for a very clear answer! I have one more question, would you use `mp_map_elem_t` or `mp_rom_map_elem_t` for the locals dict table?

**Reply by dlech** [reply]
Locals dicts are usually `mp_rom_map_elem_t` too.
